pdns/recursordist/negcache.cc

```diff
@@ -0,0 +1,155 @@
+/*
+ * This file is part of PowerDNS or dnsdist.
+ * Copyright -- PowerDNS.COM B.V. and its contributors
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of version 2 of the GNU General Public License as
+ * published by the Free Software Foundation.
+ *
+ * In addition, for the avoidance of any doubt, permission is granted to
+ * link this program with OpenSSL and to (re)distribute the binaries
+ * produced as the result of such linking.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+#include "negcache.hh"
+#include "misc.hh"
+#include "cachecleaner.hh"
+
+/*!
+ * Set ne to the NegCacheEntry for the last label in qname and return true
+ *
+ * \param qname    The name to look up (only the last label is used)
+ * \param now      A timeval with the current time, to check if an entry is expired
+ * \param ne       A NegCacheEntry that is filled when there is a cache entry
+ * \return         true if ne was filled out, false otherwise
+ */
+bool NegCache::getRootNXTrust(const DNSName& qname, const struct timeval& now, NegCacheEntry& ne) {
+  // An 'ENT' QType entry, used as "whole name" in the neg-cache context.
+  static const QType qtnull(0);
+  pair<negcache_t::const_iterator, negcache_t::const_iterator> range;
```

Comment:
hrm, come to think of it. We should actually iterate over it and check `d_auth.isRoot() && d_type == QType(0)` (whole name, not just some record). This in case the root denies com.|A, we don't want to deny .com 👍